src/server/hook_func.c

```diff
@@ -3857,13 +3857,11 @@ process_hooks(struct batch_request *preq, char *hook_msg, size_t msg_len,
 
 		jobid = ((struct rq_runjob *)(req_ptr.rq_run))->rq_jid;
 		t = is_job_array(jobid);
-		if (t == IS_ARRAY_NO) {
-			pjob = find_job(jobid); /* regular job */
-		} else if ((t == IS_ARRAY_Single) || (t == IS_ARRAY_Range)) {
-			pjob = find_arrayparent(jobid); /* subjob(s) */
+		if ((t == IS_ARRAY_Single) || (t == IS_ARRAY_NO)) {
+			pjob = find_job(jobid); /* regular job and single subjob */
 		}
 
-		/* an array job will fall through with pjob set to NULL */
+		/* an array job or range of subjobs will fall through with pjob set to NULL */
```

Comment:
Overall, the change looks okay. The only thing is that getting a range of subjobs will now fall through and get a NULL pjob causing runjob hook rejection. Intuitively, I don't think the code will get into this situation since a range of subjobs being run would just break down to individual subjobs. But this still need to be verified. With this fix, try running this simple test:
With a runjob hook event in place,
Turn off scheduling
Submit a job array, ex. qsub -J 1-4
Run the job: qrun -H <hostname> 2-4

Make sure subjobs are run with each successfully executing runjob hook.

Reply:
yes, runjob event wont make sense to anything other than a regular job or a single subjob.
I verified as suggested.  Pasting the output below
```
[shrini@dev939 pbspro]$ sudo /opt/pbs/bin/qmgr -c "s s scheduling = False"
[shrini@dev939 pbspro]$ qstat -Bf | grep scheduling
    scheduling = False
[shrini@dev939 pbspro]$ qsub -J 1-4 -- /bin/sleep 10
1[].dev939
[shrini@dev939 pbspro]$ qstat -t
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
1[].dev939        STDIN            shrini                   0 Q workq
1[1].dev939       STDIN            shrini                   0 Q workq
1[2].dev939       STDIN            shrini                   0 Q workq
1[3].dev939       STDIN            shrini                   0 Q workq
1[4].dev939       STDIN            shrini                   0 Q workq
[shrini@dev939 pbspro]$ sudo /opt/pbs/bin/qrun -H - 1[2-4]
[shrini@dev939 pbspro]$ qstat -t
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
1[].dev939        STDIN            shrini                   0 B workq
1[1].dev939       STDIN            shrini                   0 Q workq
1[2].dev939       STDIN            shrini            00:00:00 R workq
1[3].dev939       STDIN            shrini            00:00:00 R workq
1[4].dev939       STDIN            shrini            00:00:00 R workq
[shrini@dev939 pbspro]$ qstat -t
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
1[].dev939        STDIN            shrini                   0 B workq
1[1].dev939       STDIN            shrini                   0 Q workq
1[2].dev939       STDIN            shrini                   0 X workq
1[3].dev939       STDIN            shrini                   0 X workq
1[4].dev939       STDIN            shrini                   0 X workq

[shrini@dev939 pbspro]$ sudo /opt/pbs/bin/tracejob 1[1] | grep site

tracejob: Couldn't find Job Id 1[1].dev939 in logs of past 1 day

[shrini@dev939 pbspro]$ sudo /opt/pbs/bin/tracejob 1[2] | grep site
01/31/2019 11:36:54  S    'runjob_hook' hook set job's Resource_List.site = site_value
01/31/2019 11:36:54  A    user=shrini group=shrini project=_pbs_project_default jobname=STDIN queue=workq ctime=1548914467 qtime=1548914467 etime=1548914467 start=1548914814 exec_host=dev939/0 exec_vnode=(dev939:ncpus=1) Resource_List.ncpus=1 Resource_List.nodect=1 Resource_List.place=pack Resource_List.select=1:ncpus=1 Resource_List.site=site_value resource_assigned.ncpus=1
01/31/2019 11:37:04  A    user=shrini group=shrini project=_pbs_project_default jobname=STDIN queue=workq ctime=1548914467 qtime=1548914467 etime=1548914467 start=1548914814 exec_host=dev939/0 exec_vnode=(dev939:ncpus=1) Resource_List.ncpus=1 Resource_List.nodect=1 Resource_List.place=pack Resource_List.select=1:ncpus=1 Resource_List.site=site_value session=120872 end=1548914824 Exit_status=0 resources_used.cpupercent=0 resources_used.cput=00:00:00 resources_used.mem=0kb resources_used.ncpus=1 resources_used.vmem=0kb resources_used.walltime=00:00:10 run_count=1
[shrini@dev939 pbspro]$ sudo /opt/pbs/bin/tracejob 1[3] | grep site
01/31/2019 11:36:54  S    'runjob_hook' hook set job's Resource_List.site = site_value
01/31/2019 11:36:54  A    user=shrini group=shrini project=_pbs_project_default jobname=STDIN queue=workq ctime=1548914467 qtime=1548914467 etime=1548914467 start=1548914814 exec_host=dev939/1 exec_vnode=(dev939:ncpus=1) Resource_List.ncpus=1 Resource_List.nodect=1 Resource_List.place=pack Resource_List.select=1:ncpus=1 Resource_List.site=site_value resource_assigned.ncpus=1
01/31/2019 11:37:05  A    user=shrini group=shrini project=_pbs_project_default jobname=STDIN queue=workq ctime=1548914467 qtime=1548914467 etime=1548914467 start=1548914814 exec_host=dev939/1 exec_vnode=(dev939:ncpus=1) Resource_List.ncpus=1 Resource_List.nodect=1 Resource_List.place=pack Resource_List.select=1:ncpus=1 Resource_List.site=site_value session=120873 end=1548914825 Exit_status=0 resources_used.cpupercent=0 resources_used.cput=00:00:00 resources_used.mem=0kb resources_used.ncpus=1 resources_used.vmem=0kb resources_used.walltime=00:00:10 run_count=1
[shrini@dev939 pbspro]$ sudo /opt/pbs/bin/tracejob 1[4] | grep site
01/31/2019 11:36:54  S    'runjob_hook' hook set job's Resource_List.site = site_value
01/31/2019 11:36:54  A    user=shrini group=shrini project=_pbs_project_default jobname=STDIN queue=workq ctime=1548914467 qtime=1548914467 etime=1548914467 start=1548914814 exec_host=dev939/2 exec_vnode=(dev939:ncpus=1) Resource_List.ncpus=1 Resource_List.nodect=1 Resource_List.place=pack Resource_List.select=1:ncpus=1 Resource_List.site=site_value resource_assigned.ncpus=1
01/31/2019 11:37:05  A    user=shrini group=shrini project=_pbs_project_default jobname=STDIN queue=workq ctime=1548914467 qtime=1548914467 etime=1548914467 start=1548914814 exec_host=dev939/2 exec_vnode=(dev939:ncpus=1) Resource_List.ncpus=1 Resource_List.nodect=1 Resource_List.place=pack Resource_List.select=1:ncpus=1 Resource_List.site=site_value session=120874 end=1548914825 Exit_status=0 resources_used.cpupercent=0 resources_used.cput=00:00:00 resources_used.mem=0kb resources_used.ncpus=1 resources_used.vmem=0kb resources_used.walltime=00:00:10 run_count=1
[shrini@dev939 pbspro]$
```